src/tools/rest-backend/authentication_application.cpp

```diff
@@ -0,0 +1,307 @@
+#include <ctime>
+#include <iostream>
+#include <regex>
+#include <stdlib.h>
+#include <string.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <jwt/jwt.h>
+#include <openssl/sha.h>
+
+#include "authentication_application.hpp"
+#include "crypto.hpp"
+#include "root_application.hpp"
+#include "service.hpp"
+#include "user_application.hpp"
+
+namespace kdbrest
+{
+
+/**
+ * @brief the constructor fot the authentication endopint application.
+ * @param srv a service container
+ */
+AuthenticationApp::AuthenticationApp (cppcms::service & srv) : cppcms::application (srv)
+{
+
+	dispatcher ().assign ("", &AuthenticationApp::authenticate, this);
+	mapper ().assign ("");
+}
+
+/**
+ * @brief handler for the authentication resource.
+ *		  attempts an authentication with the submitted credentials. in case the
+ *		  attempt succeeds, a session token will be returned, otherwise an error.
+ */
+void AuthenticationApp::authenticate ()
+{
+	RootApp::setCORSHeaders (response (), "POST,OPTIONS");
+
+	if (request ().request_method () == "POST")
+	{
+		// check if request data is of type application/json
+		if (request ().content_type_parsed ().media_type () != "application/json")
+		{
+			RootApp::setNotAcceptable (response (), "You have supplied an unsupported Content-Type.",
+						   "REQUEST_UNSUPPORTED_CONTENT_TYPE");
+			return;
+		}
+
+		// try to parse request body data
+		cppcms::json::value requestData;
+		try
+		{
+			requestData = RootApp::parsePostDataAsJson (request ());
+		}
+		catch (kdbrest::exception::InvalidPostDataFormatException & e)
+		{
+			RootApp::setBadRequest (response (), "The submitted data is not of type application/json.",
+						"REQUEST_MALFORMED_DATA");
+			return;
+		}
+
+		// required request fields
+		std::string username;
+		std::string password;
+
+		// find username
+		try
+		{
+			username = requestData.get<std::string> ("username");
```

Comment:
I changed the retrieval of inputs to something different in places where it makes sense (3 or more input fields), but for two fields it isn't viable, so I'd like to keep it that way here.